chemprop/cli/train.py

```diff
@@ -542,170 +531,220 @@ def main(args):
+                if len(test_data[0]) > 0:
+                    test_dsets = [make_dataset(data, args.rxn_mode) for data in test_data[fold_idx]]
+                    test_dset = data.MulticomponentDataset(test_dsets)
+                    test_loader = data.MolGraphDataLoader(
+                        test_dset, args.batch_size, args.num_workers, shuffle=False
+                    )
+                else:
+                    test_loader = None
+            else:
+                if len(test_data) > 0:
+                    test_dset = make_dataset(test_data[fold_idx], args.rxn_mode)
+                    test_loader = data.MolGraphDataLoader(
+                        test_dset, args.batch_size, args.num_workers, shuffle=False
+                    )
+                else:
+                    test_loader = None
+
+            mpnn_cls = MulticomponentMPNN if multicomponent else MPNN
+            model = mpnn_cls(
+                mp_block,
+                agg,
+                predictor_ffn,
+                True,
+                None,
+                args.task_weights,
+                args.warmup_epochs,
+                args.init_lr,
+                args.max_lr,
+                args.final_lr,
+            )
+            logger.info(model)
+
+            monitor_mode = "min" if model.metrics[0].minimize else "max"
+            logger.debug(f"Evaluation metric: '{model.metrics[0].alias}', mode: '{monitor_mode}'")
+
+            if no_cv:
+                output_dir = args.output_dir
+            else:
+                output_dir = args.output_dir / f"fold_{fold_idx}"
+
+            tb_logger = TensorBoardLogger(output_dir / "tb_logs")
+            checkpointing = ModelCheckpoint(
+                output_dir / "chkpts",
+                "{epoch}-{val_loss:.2f}",
+                "val_loss",
+                mode=monitor_mode,
+                save_last=True,
+            )
+
+            early_stopping = EarlyStopping("val_loss", patience=5, mode=monitor_mode)
+
+            trainer = pl.Trainer(
+                logger=tb_logger,
+                enable_progress_bar=True,
+                accelerator="auto",
+                devices=args.n_gpu if torch.cuda.is_available() else 1,
+                max_epochs=args.epochs,
+                callbacks=[checkpointing, early_stopping],
+            )
+            trainer.fit(model, train_loader, val_loader)
+
+            if test_loader is not None:
+                if args.task_type == "regression":
+                    model.predictor.register_buffer("loc", torch.tensor(scaler.mean_).view(-1, 1))
+                    model.predictor.register_buffer(
+                        "scale", torch.tensor(scaler.scale_).view(-1, 1)
+                    )
+                results = trainer.test(model, test_loader)[0]
+                logger.info(f"Test results: {results}")
+
+            p_model = args.output_dir / "model.pt"
+            input_scalers = []  # TODO: we should add descriptor scalers here
```

Comment:
Again, consider breaking this TODO into an issue

Reply:
(and removing the inline comment)

Reply:
Added this issue, will remove the inline comment: https://github.com/chemprop/chemprop/issues/640